armsrc/lfops.c

```diff
@@ -1220,6 +1222,265 @@ void T55xxWriteBit(int bit) {
 	WaitUS(WRITE_GAP);
 }
 
+void T55xxWrite_LLR (void)
+{
+	TurnReadLFOn (Reference_llr);
+	FpgaWriteConfWord(FPGA_MAJOR_MODE_OFF);
+	WaitUS(WRITE_GAP);	
+}
+
+#define START_GAPlz 31*8 
+#define WRITE_GAPlz 20*8 
+#define WRITElz_0   18*8 
+#define WRITElz_1   40*8 
+#define READ_GAP  15*8 
+
+void T55xxWriteBit_Leading0(int bit) {
+	if (!bit)
+		TurnReadLFOn(WRITElz_0);
+	else
+		TurnReadLFOn(WRITElz_1);
+	FpgaWriteConfWord(FPGA_MAJOR_MODE_OFF);
+	WaitUS(WRITE_GAPlz);
+//	WaitUS(160);
+}
+
+#define START_GAP1of4 31*8 // SPEC:  1*8 to 50*8 - typ 10*8 (or 15fc)
+#define WRITE_GAP1of4 20*8 // SPEC:  1*8 to 20*8 - typ 10*8 (or 10fc)
+//  00 = reference			// 8 * 8 -  - 68 * 8
+#define WRITE1of4_00  18*8 // SPEC:  8*8 to 68*8 - typ 24*8 (or 24fc)
+#define WRITE1of4_01  34*8 // SPEC: dref+9  - dref+16 - dref+24 
+#define WRITE1of4_10  50*8 // SPEC: dref+25 - dref+32 - dref+40
+#define WRITE1of4_11  66*8 // SPEC: dref+41 - dref+48 - dref+56
+#define READ1of4_GAP  15*8 
+
+void T55xxWriteBit_1of4(int bits) {
+	
+	switch (bits)
+	{
+		case 0 :  TurnReadLFOn(WRITE1of4_00); break;
+		case 1 :  TurnReadLFOn(WRITE1of4_01); break;
+		case 2 :  TurnReadLFOn(WRITE1of4_10); break;
+		case 3 :  TurnReadLFOn(WRITE1of4_11); break;
+		default:
+			TurnReadLFOn(WRITE1of4_00);
+	}		
+
+	FpgaWriteConfWord(FPGA_MAJOR_MODE_OFF);
+	WaitUS(WRITE_GAP1of4);
+//	WaitUS(160);
+}
+
+void T55xxWriteBlockExt_Leading0 (uint32_t Data, uint32_t Block, uint32_t Pwd, uint8_t arg) {
+
+	LED_A_ON();
+	bool PwdMode = arg & 0x1;
+	uint8_t Page = (arg & 0x2)>>1;
+	bool testMode = arg & 0x4;
+	uint32_t i = 0;
+	
+	// Set up FPGA, 125kHz
+	LFSetupFPGAForADC(95, true);
+	StartTicks();
+	// make sure tag is fully powered up...
+	WaitMS(5);
+	// Trigger T55x7 in mode.
+	FpgaWriteConfWord(FPGA_MAJOR_MODE_OFF);
+
+	WaitUS(START_GAPlz);
+	
+	
+	/* 
+	   0 	: Leading Zero
+	   11   : Opcode
+	   00   : Fixed 00 if protected write (i.e. have password)
+	   <32 bit Password>
+	   0 	: Lock Bit
+	   <32 bit data> 
+	   <3 bit addr>
+
+	   	Standard Write : 0 1p L <32 data bits> <3 bit addr>
+		                 0 10 0 00000000000000000000000000000000 001
+		Protected Write: 0 1p 00 <32 pwd bits> L <32 data bits> <3 bit addr>
+					     0 10 00 00000000000000000000000000000000 0 00000000000000000000000000000000 001
+		Wake Up			 0 10 00 <32 pwd bits>
+		Protected Read   0 1p 00 <32 pwd bits> 0 <3 bit addr>
+		Standard Read    0 1p 0 <3 bit addr>
+		Page 0/1 read    0 1p
+		Reset            0 00
+		
+	*/
+		T55xxWriteBit_Leading0 (0); //T55xxWriteBit(0);
+				
+	
+		if (testMode) Dbprintf("TestMODE");
```

Comment:
Doesn't `Dbprintf()` take too long? And therefore spoils the timing? Or isn't this an issue in lf?

Reply:
I assume here you mean put the Leading0 write  after the dbprintf. (the actual Dbprintf("TestMODE") was already there.
I will move the initial leading 0 write bit call as needed.

Reply:
Yes. This is not your code, but I noticed it while reviewing. The Dbprintf() will add more time (too much?) to the gap than intended. It therefore should come before the T55xxWriteBit_Leading0() where it doesn't hurt. Same in all other T55xWriteBit functions. BTW: this is an example of the downside of code duplication, you need to remember to apply the same fix in all cloned functions,

Reply:
100% Agree.  I just missed that, already moved in dev :)